evm/src/executor/inspector/cheatcodes/ext.rs

```diff
@@ -160,21 +159,46 @@ fn get_env(key: &str, r#type: ParamType, delim: Option<&str>) -> Result<Bytes, B
         .map_err(|e| e.into())
 }
 
-fn read_file(path: &str) -> Result<Bytes, Bytes> {
-    let data = fs::read_to_string(path).map_err(|err| err.to_string().encode())?;
+macro_rules! check_path_allowed {
+    ($state:ident, $absolute_path:ident) => {
+        // Check if any of the allowed paths is the prefix for our absolute path.
+        if !$state
+            .config
+            .allowed_paths
+            .iter()
+            .any(|allowed_path| $absolute_path.starts_with(allowed_path))
+        {
+            return Err(util::encode_error("Path is not allowed."))
+        }
+    };
+}
+
+fn full_path(state: &mut Cheatcodes, path: impl AsRef<Path>) -> PathBuf {
+    state.config.root.join(path)
+}
+
+fn read_file(state: &mut Cheatcodes, path: impl AsRef<Path>) -> Result<Bytes, Bytes> {
+    let path = full_path(state, &path);
+    check_path_allowed!(state, path);
+
+    let data = fs::read_to_string(path).map_err(util::encode_error)?;
 
     Ok(abi::encode(&[Token::String(data)]).into())
 }
 
-fn read_line(state: &mut Cheatcodes, path: &str) -> Result<Bytes, Bytes> {
+fn read_line(state: &mut Cheatcodes, path: impl AsRef<Path>) -> Result<Bytes, Bytes> {
+    let path = full_path(state, &path);
+    check_path_allowed!(state, path);
+
     // Get reader for previously opened file to continue reading OR initialize new reader
-    let reader =
-        state.context.opened_read_files.entry(path.to_string()).or_insert(BufReader::new(
-            fs::File::open(path).map_err(|err| err.to_string().encode())?,
-        ));
+    let reader = state
+        .context
+        .opened_read_files
+        .entry(path.clone())
+        .or_insert(BufReader::new(fs::File::open(&path).map_err(util::encode_error)?));
```

Comment:
I believe `foundry_common::fs::open` is missing, mind adding this and a corresponding Error Variant (Open) to FsPathError ?

Reply:
yep, also added `create_file`. Only `std::fs` usage in this file that's left is https://github.com/foundry-rs/foundry/blob/d0cf050b89258c09db55eaef98a61e0e48dc284d/evm/src/executor/inspector/cheatcodes/ext.rs#L213-L217
but not sure whether we need to incorporate this stuff into `foundry_common::fs`

Reply:
I think this is fine, since the flexibility to configure OpenOptions here is needed